### PR TITLE
Made "via convoy" and "own fleet convoys" treated the same.

### PR DESCRIPTION
**Pull request by zond**
See #128.

Right now, if a unit that can move over land is ordered to move via convoy, but the convoy fails for any reason, the unit tries to move over land.

If a unit that can move over land is convoyed by a fleet of the same player, and the convoy fails for any reason, the unit will _not_ try to move over land.

This is unintuitive and like the author of #128 I don't like it.

Which brings us to this PR, which makes both cases work the same - army moves over land if convoy is unavailable.

@tttppp, @jensforsgard, please provide input.

### CODE REVIEW

**Comment by tttppp** [comment]
Relevant discussion from Discord:

> **zond**
> If the same player that ordered the move also orders a convoy, then that is interpreted the same way as 'via convoy'.
> The main change here in this PR is that we don't assume we can pick the land route instead of the convoy route just because the convoy is dislodged (in the via convoy case)
> Which means, that a 'via convoy' is only respected if there is a matching convoy path, and if that convoy path is dislodged then it's treated the same way as if the unit owner tried to convoy but was dislodged: i.e. a missing convoy path error
> 
> **tttppp**
> > If the same player that ordered the move also orders a convoy, then that is interpreted the same way as 'via convoy'.
>
> Any fleet they own agrees with the convoy?
> Or they have to supply a complete convoy path?
>
> **zond**
> Any fleet they have, that could be on a convoy path, has a matching convoy order
> They don't need a complete path, but the fleet has to be a possible part of a path

File: variants/classical/classical_test.go [hunk 1]
```diff
@@ -745,3 +745,78 @@ func TestMessages(t *testing.T) {
...
+	judge.SetUnit("tys", godip.Unit{godip.Fleet, godip.England})
+	judge.SetUnit("wes", godip.Unit{godip.Fleet, godip.France})
+	judge.SetUnit("ion", godip.Unit{godip.Fleet, godip.France})
+	judge.SetOrder("nap", orders.Move("nap", "rom"))
+	judge.SetOrder("tys", orders.Convoy("tys", "nap", "rom"))
+	judge.SetOrder("wes", orders.Move("wes", "tys"))
+	judge.SetOrder("ion", orders.SupportMove("wes", "wes", "tys"))
+	judge.Next()
+	if found := judge.Resolutions()["nap"]; found != nil {
+		t.Errorf("Wanted failure for nap, got %v", found)
```
Comment (tttppp) [inline variants/classical/classical_test.go:786]: Wanted success for nap

**Review by tttppp — COMMENTED** [review]
This looks good to me. The fix is magic anyway.

I think it would be possible to get one or two extra test cases using:
* England Army Wales move to London (not "via convoy")
* England Fleet English Channel (e.g. not ordering a convoy)
* England Fleet Norwegian Sea Convoys Wales to London
* France Fleets Irish Sea, NAO, North Sea
* Germany Fleets SKA and HEL breaking convoy in North Sea

In this way we would be able to test that England order a single fleet that could be involved in a convoy causes the convoy to be prioritised above the normal move (even if they don't convoy the sensible short way).

File: variants/classical/classical_test.go (hunk 1 above)
Comment (zond) [reply variants/classical/classical_test.go]: Thanks!

**Comment by zond** [comment]
> This looks good to me. The fix is magic anyway.

Not magic! The function that checks if the army _must_ convoy used to check:

- Does the army not have a land path?
- Does the army say 'ViaConvoy' _and_ have a working (as in not dislodged) convoy path?
- Does the owner of the army issue a convoy order for the same move?

If one was true, it had to convoy.

Now the second criterion is changed to

- Does the army say 'ViaConvoy' _and_ have a convoy path where the fleets are actually ordered to convoy the same move?

**Review by tttppp — APPROVED** [review]
LGTM!

**Comment by zond** [comment]
Copied from an email discussing this:
```
Consider two typical convoy paradoxes (I'm using the term paradox to describe situations that doesn't have a simple solution using your adjudication algorithm):

A) Convoy order that would result in dislodgment of the convoying fleet if it failed.
B) Convoy order that would result in dislodgment of the convoying fleet if it succeeded.

They are paradoxes since in case A) both results are consistent, and in case B) none of them are.

But 6.G.11 doesn't fit this pattern:

If the convoy succeeds, then the support for the dislodgment of the convoy is broken, and the convoy can still succeed, which is consistent.

If the convoy fails, then the army moves via land and breaks the support, which lets the convoy succeed, which is inconsistent.

I.e., exactly one case is consistent, and there is no paradox, and the convoy can proceed.

On the other hand, if we assume that 'intent to convoy' forbids land movement (which many automatic adjudicators including mine do - possibly in error, which is what brought this up), then a failed convoy leads to no support broken, and a dislodged convoy, and also a consistent result - i.e. a paradox due to two consistent alternatives.

So, according to this analysis, 6.G.11 is a paradox only if intent to convoy forbids land movement. Is this intended (forgive the pun)?

I interpret "BOTH A CONVOY ROUTE AND AN OVERLAND ROUTE. If an army could arrive at its destination either overland or by convoy, one route must be considered and the other disregarded, depending upon intent as shown by the totality of the orders written by the player governing the army" as "land route is not possible when there is a convoy intent", but that's not the most recent rules, and possibly not the rule people today expect.
```

**Comment by zond** [comment]
Based on the discussion in the email above, and the reasonableness of letting armies take land route when convoy route is unavailable, I now reversed the behavior so that it works like that instead.

**Comment by zond** [comment]
@tttppp PTAL again now that I reversed the behavior. Still looking good?

**Comment by tttppp** [comment]
I think there's a good level of testing here and the logic seems sensible to me. In addition this whole thing is an edge case if none of the tests failed. Did you try running against the Droidippy games?

Can we change the DATC expected result and still claim to be DATC compliant?

**Comment by tttppp** [comment]
Although not completely an edge case if that Empires and Coalitions test game hit it!

**Comment by zond** [comment]
> I think there's a good level of testing here and the logic seems sensible to me. In addition this whole thing is an edge case if none of the tests failed. Did you try running against the Droidippy games?

I ran it against a good number of them, but not all IIRC.

> Can we change the DATC expected result and still claim to be DATC compliant?

DATC is a bunch of test cases and some of them (like this one) doesn't have a hard and fast correct response but more like alternatives and recommendations. As long as we cover the cases in a sensible way I think it's as intended.

**Comment by zond** [comment]
> Although not completely an edge case if that Empires and Coalitions test game hit it!

But that was random moves, right?

Also - I was unable to understand why it failed! I think the changed behavior in that particular game before and after this patch is probably a quite exotic bug.

What I saw was that a convoy and a move bounced differently due to this change.. and I don't understand why.

I should probably try to figure out what the hell happened there for completeness sake.

**Comment by zond** [comment]
Discussion in Discord made me realize that godip hasn't implemented the rule "convoyed armies don't break support against the convoying fleets", which makes 6.G.11 into a paradox again, and makes the "convoy intent blocks land movement vs it doesn't" a more arbitrary decision.

What do we want? What is the correct rule for "convoy intent blocks land movement vs it doesn't"?

**Comment by TimothyJones** [comment]
To summarise my discord position - the rulebook says:

> In some rare cases, orders are written so that an Army could arrive
at its destination either by land or convoy. When this happens, the
following qualifiers apply:
> • If at least one of the convoying Fleets belongs to the player
 who controls the Army, then the convoy is used. The land
 route is disregarded.
> • If none of the convoying Fleets belongs to the player who
 controls the Army, then the land route is used. However, the
 player controlling the Army can use the convoy route if he
 or she indicated “via convoy” on the Army move order
 in question. 
 
 I believe this means that if the convoy intent is expressed by ordering a) one of your fleets to make the convoy or b) specifying "via convoy" on the move order, then the convoy route is exclusively used. 

Also: This PR removes most of variants/empiresandcoalitions/games/game_1.txt, is it meant to do that?

Also also: The code change reads well, and I don't even know the rest of the code.

**Comment by zond** [comment]
> Also: This PR removes most of variants/empiresandcoalitions/games/game_1.txt, is it meant to do that?

Yes - some weird thing happened in that test, and I didn't think it was actually related to this. I created https://github.com/zond/godip/issues/131 to track that fresh hell.

> Also also: The code change reads well, and I don't even know the rest of the code.

**Comment by tttppp** [comment]
> Although not completely an edge case if that Empires and Coalitions test game hit it!
>
> But that was random moves, right?
Yes, that game was random moves.

**Comment by zond** [comment]
Alright, I created https://github.com/zond/godip/pull/133 that implements a missing rule in the adjudicator.

After that one is landed, we should consider how we want this solved.

The main question is: Do armies try the land route even if a convoy intent is clear?

**Comment by zond** [comment]
Now that #134 has been merged, this is superseded and I'll close it.
